src/main/java/com/google/cloud/pubsublite/kafka/SinglePartitionSubscriber.java

```diff
@@ -0,0 +1,150 @@
+    this.subscriber =
+        subscriberFactory.newPullSubscriber(partition, initialSeek, this::onSubscriberReset);
+  }
+
+  /** Executes a client-initiated seek. */
+  void clientSeek(SeekRequest request) throws CheckedApiException {
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      subscriber.close();
+      subscriber = subscriberFactory.newPullSubscriber(partition, request, this::onSubscriberReset);
+    }
+  }
+
+  ApiFuture<Void> onData() {
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      return subscriber.onData();
+    }
+  }
+
+  @GuardedBy("monitor.monitor")
+  private ArrayDeque<SequencedMessage> pullMessages() throws CheckedApiException {
+    ArrayDeque<SequencedMessage> messages = new ArrayDeque<>();
+    for (Optional<SequencedMessage> message = subscriber.messageIfAvailable();
+        message.isPresent();
+        message = subscriber.messageIfAvailable()) {
+      messages.add(message.get());
+    }
+    return messages;
+  }
+
+  /** Pulls all available messages. */
+  ArrayDeque<SequencedMessage> getMessages() throws CheckedApiException {
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      ArrayDeque<SequencedMessage> messages = pullMessages();
+      if (!messages.isEmpty()) {
+        lastReceived = Optional.of(Iterables.getLast(messages).offset());
+        needsCommitting = true;
+      }
+      return messages;
+    }
+  }
+
+  Optional<Long> position() {
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      return lastReceived.map(lastReceived -> lastReceived.value() + 1);
+    }
+  }
+
+  /** Executes a client-initiated commit. */
+  ApiFuture<Void> commitOffset(Offset offset) {
+    return committer.commitOffset(offset);
+  }
+
+  /** Auto-commits the offset of the last received message. */
+  Optional<ApiFuture<Offset>> autoCommit() {
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      if (!needsCommitting) return Optional.empty();
+      checkState(lastReceived.isPresent());
+      needsCommitting = false;
+      // The Pub/Sub Lite commit offset is one more than the last received.
+      Offset toCommit = Offset.of(lastReceived.get().value() + 1);
+      return Optional.of(
+          ApiFutures.transform(
+              committer.commitOffset(toCommit),
+              ignored -> toCommit,
+              MoreExecutors.directExecutor()));
+    }
+  }
+
+  void close() throws CheckedApiException {
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      subscriber.close();
+    }
+    committer.stopAsync().awaitTerminated();
+  }
+
+  private boolean onSubscriberReset() throws CheckedApiException {
+    // Handle an out-of-band seek notification from the server. There must be no pending commits
+    // after this function returns.
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      // Discard undelivered messages.
+      pullMessages();
+      // Prevent further auto-commits until post-seek messages are received.
+      needsCommitting = false;
```

Comment:
I think you should also reset lastReceived

Reply:
I think we shouldn't reset it here, as [PubsubLiteConsumer.position()](https://github.com/googleapis/java-pubsublite-kafka/blob/783b3fcb19863509d5ca60f6bd558e5235c15885/src/main/java/com/google/cloud/pubsublite/kafka/PubsubLiteConsumer.java#L395) will retrieve committed offsets from the server, if `lastReceived` is empty. This would be incorrect if the committer has yet to flush the last commit.

Setting `needsCommittering = false` was enough to prevent further commits.